src/cache/cache_set.py

```diff
@@ -0,0 +1,275 @@
+        return self.mesi_state.bits
+
+    def __str__(self) -> str:
+        """String representation of the cache line"""
+        return (
+            f"Tag={hex(self.tag) if self.tag is not None else 'None'}, "
+            f"State={self.mesi_state.name}, "
+            f"In L1={self.in_l1}"
+        )
+
+
+class CacheSetPLRUMESI:
+    """
+    Implements an N-way set associative cache set with MESI protocol and PLRU replacement.
+    Default configuration is 16-way, with supported ways from 1 to 32 (must be power of 2).
+    """
+
+    MAX_WAYS = 32  # Maximum supported ways
+    MIN_WAYS = 1  # Minimum supported ways
+
+    def __init__(self, num_ways: int = 16):
+        # Validate num_ways
+        if not isinstance(num_ways, int):
+            raise TypeError("Number of ways must be an integer")
+        if num_ways < self.MIN_WAYS or num_ways > self.MAX_WAYS:
+            raise ValueError(
+                f"Number of ways must be between {self.MIN_WAYS} and {self.MAX_WAYS}"
+            )
+        if not (num_ways & (num_ways - 1) == 0):
+            raise ValueError("Number of ways must be a power of 2")
+
+        self.num_ways = num_ways
+        self.tree_levels = int(math.log2(num_ways))  # Levels in PLRU tree
+
+        # Initialize cache lines
+        self.ways = [CacheLine() for _ in range(num_ways)]
+
+        # Initialize PLRU state - only need (num_ways - 1) bits for tree
+        self.state = 0
+
+    def __get_parent(self, node: int) -> int:
+        """Helper method for PLRU tree traversal"""
+        return (node - 1) // 2
+
+    def __update_plru(self, way_index: int) -> None:
+        """
+        Update PLRU state when a way is accessed. If left child is accessed, parent bit is cleared.
+        If right child is accessed, parent bit is set.
+        """
+        if way_index >= self.num_ways:
+            raise ValueError(
+                f"Way index {way_index} out of range for {self.num_ways}-way set"
+            )
+
+        # PLRU bits are updated with a bottom up tree traversal
+        leaf_node = way_index + (
+            self.num_ways - 1
+        )  # Convert way index to a leaf_node number
+        node = self.__get_parent(
+            leaf_node
+        )  # Use leaf node to get first internal node (i.e. PLRU bit)
+
+        if way_index % 2 == 0:  # Left child accessed
+            self.state &= ~(1 << node)  # Clear node's bit
+        else:  # Right child accessed
+            self.state |= 1 << node  # Set node's bit
+
+        while node > 0:
+            parent = self.__get_parent(node)
+            if node == 2 * parent + 1:  # Current node is left child
+                self.state &= ~(1 << parent)  # Clear parent's bit
+            else:  # Current node is right child
+                self.state |= 1 << parent  # Set parent's bit
+            node = parent
+
+    def __get_plru_victim(self) -> int:
+        """
+        Find the way to replace according to PLRU policy.
```

Comment:
I think it could be helpful to be explicit what the PLRU policy is or how it works in this method. Specifically that to find the PLRU node, you traverse the tree and go left when the node value is 1, and you go right when the node value is 0. Similar to the above, I think the code is clear, but having a basic description jump-starts understanding when revisiting the logic.